### ISSUE TITLE
[Newtyper] Add support for list operations

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#215]
I have added the following to `Codegen`:

```
   case Expression.FNil(tpe, loc) =>
      // Call the static method ca.uwaterloo.flix.runtime.value.FNil.get().
      ???

    case Expression.FList(hd, tl, tpe, loc) =>
      // Generate code for hd and tl.
      // Call the constructor of ca.uwaterloo.flix.runtime.value.FList(hd, tl).
      ???

    case Expression.IsNil(exp, loc) =>
      // Generate code for exp.
      // Perform INSTANCEOF for ca.uwaterloo.flix.runtime.value.FNil.
      ???

    case Expression.IsList(exp, loc) =>
      // Generate code for exp.
      // Perform INSTANCEOF for ca.uwaterloo.flix.runtime.value.FList.
      ???

    case Expression.GetHead(exp, tpe, loc) =>
      // Generate code for exp.
      // Cast the result to ca.uwaterloo.flix.runtime.value.FList
      // Call the INSTANCE method ca.uwaterloo.flix.runtime.value.FList.getHd
      ???

    case Expression.GetTail(exp, tpe, loc) =>
      // Generate code for exp.
      // Cast the result to ca.uwaterloo.flix.runtime.value.FList
      // Call the INSTANCE method ca.uwaterloo.flix.runtime.value.FList.getTl
      ???
```

@mhyee if you have the time, could you implement these operations for me? 

Note that I created two Java classes: `FNil` and `FList` which use Java conventions, hopefully this should both make Codegen easier and help with interop (down the road).

**Comment by magnus-madsen** [comment]
I am hoping that by implementing this, we can discover where exactly the bug is in  #214.

**Comment by mhyee** [comment]
I'm going to punt this for now, since it seems like a rather large chunk of work.

**Comment by magnus-madsen** [comment]
I added some code, but there are still issues with boxing. I don't know yet if they are deep issues, or if I just need to make a small fix.

**Comment by magnus-madsen** [comment]
The plan is to remove list support from the compiler (and rely on the enum/tag mechanism). The use of `::` is then a simple syntactic transformation applied early in the compiler.

**Comment by magnus-madsen** [comment]
I am bit stuck on this due to the nature of `List` being recursive, but I think the solution is to change the definition of `Type.Enum` to contain solely the `EnumSym` (and not the cases of the variant). Then the cases must be looked up during type inference.

**Comment by magnus-madsen** [comment]
We now have recursive polymorphic types. However, the code generation backend doesn't know how to compile a function of type e.g. `a -> a` , or `List[a] -> a`. 

@mhyee Can you give an assessment of how this is to be done?

**Comment by magnus-madsen** [comment]
NB: This problem may not have an easy answer, because there is the question of monomorphization, e.g. what is `List[Int]` compiled as?

**Comment by mhyee** [comment]
Java does type erasure, so it would generate code for `Object -> Object` and `List[Object] -> Object`. Code generation is easier and faster, but now you have to box everything at run time.

C++ will generate specialized versions of functions and classes, based on what you actually use in your code. So if you have a function `foo` of type `a -> a` and it's never called, C++ won't generate code for it. If you call `foo(42)`, C++ will deduce that you're calling `foo` on an int, so it'll generate code for `foo(x: Int): Int`.

This is more complicated, and compilation becomes slower. You also generate a lot more code, but then that code will be faster.

This is a bit similar to what we do with functional interfaces, but it will be a lot more complicated. For example, if there's a polymorphic library function, we can't generate code for it until all the user code has been looked at.

I'm not sure what Haskell or OCaml do. I think Dotty does some sort of specialization similar to what I described for C++?

**Comment by magnus-madsen** [comment]
I think it would be interesting to do monomorphization in Flix. I would expect most Flix programs to be small to medium sized, so I think increased code size will be tolerable. Apparently Mlton (an ML compiler optimized for performance) does monomorphization without too much difficulty:

https://news.ycombinator.com/item?id=7965709
http://www.mlton.org/pipermail/mlton/2001-January/018367.html

Rust does it too (in some variant, I have not been able to find any details).

I think monomorphization would also play well with the current back-end.

What are the key steps to implement this? Something like the following?

A phase Monomorpher which replaces every use of a polymorphic function with a specialized version, e.g.:

```
def f: List[Int] = map(x => x, 1 :: 2 :: Nil)
```

becomes something like:

```
def lambda$1(x: Int): Int = x + 42
def map$1(f: Int -> Int, xs: List[Int]): List[Int] = ...
def f: List[Int] = map$1(lambda$1, 1 :: 2 :: Nil)
```

I guess the lambda may already be taken care of by the lifter (at least in this case).


Where would you put the specialized functions? In the namespace of their use, or in their declaring namespace, or in some third place?

**Comment by mhyee** [comment]
Yeah, I agree that monomorphization is probably better than erasing types. It's also consistent with the other design decisions (functions return primitives instead of boxed values, functional interfaces are specialized based on the types actually used in the program).

> What are the key steps to implement this?

Before, the compiler could simply look at a function definition and generate code, without caring about where it's called. Likewise, it could look at a function call and generate a call to a method with the specified signature, and not worry about where or what the function is.

Now, the compiler will have to consider both definitions and calls. It will be a little tricky, since calls can occur before definitions. So we would probably need some phase that determines what type specializations are needed, and then rewrite generic function calls to specialized function calls.

I think we want to put specialized functions in the declaring namespace, since there may be multiple uses of the same specialization.

Generic lambdas might need to be special-cased. If you're using the generic lambda `x => x` inside a `map` function, it can be specialized right away, once it knows what type `map` needs to be specialized to. I don't think there's much a point in sharing the lambda definitions.

**Comment by magnus-madsen** [comment]
Does the following strategy work?

0. Maintain a global map from `(Symbol.DefnSym, List[Type])` from definition symbols and types.
1. Iterate through all *non-generic* functions in the program (i.e. functions without type parameters).
2. Inside each function, for every polymorphic call, lookup the function in the global map. If it does not exist instantiate and save it, and finally replace the call to call the specialised version.
3. Delete all functions with type parameters.

**Comment by mhyee** [comment]
The strategy seems reasonable, but I suspect there will be issues that will crop up.

For example, we would still need to populate the global map.

And what if a generic function calls some other generic function? We'd need to keep a worklist, initialized with non-generic functions, and any generic function that is specialized would be added to the worklist.

Also, this phase would affect which lambda expressions are actually needed, which affects the functional interfaces being generated.

**Comment by magnus-madsen** [comment]
Fixed by #259 which will be merged in shortly.
